### ISSUE TITLE
Linux : compilation error on Manjaro (and Xubuntu 20.04.1)

### ISSUE THREAD

**Issue by FredSoftwares** [vdemichev/DiaNN#68]
Hi Vadim,

I tried to compile DiaNN 1.7.12 on Linux Manjaro (up to date) without success. I don't know all the dependencies of the software, so here are my versions of glibc and gcc : 
glibc 2.32-5
gcc 10.2.0-3
It would simplify a lot the usage of the software if you could provide a statical binary for linux, since most of the users of DiaNN are (like me) not expert in C++ development. 
I also tried to compile DiaNN 1.7.12 on the PC of a colleague using Xubuntu 20.04.1 and it does not compile, the error is the same. So compilation issues are not limited to my computer and I suppose that many other linux users gave up. All the information you provide in this post should help most linux users who are not C++ developers and cannot fix the issues by themselves.

большое спасибо ;)

```
DiaNN-1.7.12/mstoolkit$ make
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/adler32.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/compress.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/crc32.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/deflate.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/inffast.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/inflate.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/infback.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/inftrees.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/trees.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/uncompr.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/zlib-1.2.8/zutil.c -c
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H  src/mzParser/MSNumpress.cpp -c -o mzp.MSNumpress.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/mzp_base64.cpp -c -o mzp.mzp_base64_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/BasicSpectrum.cpp -c -o mzp.BasicSpectrum_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/mzParser.cpp -c -o mzp.mzParser_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/RAMPface.cpp -c -o mzp.RAMPface_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/saxhandler.cpp -c -o mzp.saxhandler_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/saxmzmlhandler.cpp -c -o mzp.saxmzmlhandler_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/saxmzxmlhandler.cpp -c -o mzp.saxmzxmlhandler_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/Czran.cpp -c -o mzp.Czran_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/mz5handler.cpp -c -o mzp.mz5handler_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/mzpMz5Config.cpp -c -o mzp.mzpMz5Config_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/mzpMz5Structs.cpp -c -o mzp.mzpMz5Structs_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/BasicChromatogram.cpp -c -o mzp.BasicChromatogram_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H src/mzParser/PWIZface.cpp -c -o mzp.PWIZface_lite.o
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/MSToolkit/Spectrum.cpp -c
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/MSToolkit/MSObject.cpp -c
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/MSToolkit/mzMLWriter.cpp -c
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/MSToolkit/pepXMLWriter.cpp -c
g++ -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H -D_NOSQLITE ./src/MSToolkit/MSReader.cpp -c -o MSReaderLite.o
Dans le fichier inclus depuis ./src/MSToolkit/MSReader.cpp:21:
./include/MSReader.h:19: attention: « _NOSQLITE » redéfini
   19 | #define _NOSQLITE
      | 
<ligne-de-commande>: note: ceci est l'emplacement d’une précédente définition
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/expat-2.2.0/xmlparse.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/expat-2.2.0/xmlrole.c -c
gcc -O3 -static -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H ./src/expat-2.2.0/xmltok.c -c
ar rcs libmstoolkitlite.a adler32.o compress.o crc32.o deflate.o inffast.o inflate.o infback.o inftrees.o trees.o uncompr.o zutil.o xmlparse.o xmlrole.o xmltok.o mzp.MSNumpress.o mzp.mzp_base64_lite.o mzp.BasicSpectrum_lite.o mzp.mzParser_lite.o mzp.RAMPface_lite.o mzp.saxhandler_lite.o mzp.saxmzmlhandler_lite.o mzp.saxmzxmlhandler_lite.o mzp.Czran_lite.o mzp.mz5handler_lite.o mzp.mzpMz5Config_lite.o mzp.mzpMz5Structs_lite.o mzp.BasicChromatogram_lite.o mzp.PWIZface_lite.o Spectrum.o MSObject.o mzMLWriter.o pepXMLWriter.o MSReaderLite.o
g++ -o ../diann-linux ../src/diann.cpp -static -Wl,--whole-archive -lpthread -Wl,--no-whole-archive -std=c++17 -lstdc++fs -fpermissive -O3 -mfpmath=sse -msse2 -march=core2 -w -L. -lmstoolkitlite -I. -I./include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DGCC -DHAVE_EXPAT_CONFIG_H
../src/diann.cpp: Dans la fonction membre « void Feature::find(bool, std::vector<Mass>&, std::vector<Mass>&, std::vector<std::pair<float, int> >&, double, double, std::vector<Match>&) »:
../src/diann.cpp:2583:4: erreur: saut à l'étiquette « finish »
 2583 |    finish:
      |    ^~~~~~
../src/diann.cpp:2552:53: note:   à partir d'ici
 2552 |     if (open && iter == 1 && pos->charge != 1) goto finish;
      |                                                     ^~~~~~
../src/diann.cpp:2553:11: note:   croise l'initialisation de « std::pair<float, int>& pr »
 2553 |     auto &pr = precursors[pos->pr];
      |           ^~
../src/diann.cpp: Dans la fonction membre « void Library::max_lfq_quant(std::vector<PG>*, std::vector<Lock>*, double, bool) »:
../src/diann.cpp:5443:4: erreur: saut à l'étiquette « save »
 5443 |    save:
      |    ^~~~
../src/diann.cpp:5396:10: note:   à partir d'ici
 5396 |     goto save;
      |          ^~~~
../src/diann.cpp:5399:8: note:   croise l'initialisation de « int ni »
 5399 |    int ni = indices.size();
      |        ^~
make: *** [Makefile:30 : all] Erreur 1
```

**Comment by vdemichev** [comment]
Hi, unfortunately GCC 9 breaks backward compatibility with GCC 7, so DIA-NN's code needs some changes to compile it with GCC 9. I think this patch https://github.com/vdemichev/DiaNN/pull/65 addresses this. I will aim to look into Linux compilation issues and resolve them in the next version. For now I would suggest applying the patch, or compiling on a system with GCC 7 and then copying the binary to the target machine. The reason I don't simply release a statically compiled binary is because any software which links to glibc must comply with its very strict licensing requirements. Linking to a different C library is possible, and I might look into this in the future.

In general, I strongly recommend using Windows to run DIA-NN, the reasons include functionality, speed, easy of use and reproducibility.

Vadim

**Comment by FredSoftwares** [comment]
Hi Vadim,
Thank you for your quick answer ! I will wait for the next version to try the software on linux.
I did a try on windows, the GUI works very well, but I had some issues with diann.exe
The executable does not find the VCOMP140.DLL whereas it is present on the system. I have already installed MS visual C++ redistributable and visual studio tools. I choose C:\program files\ as target install directory because if I choose the proteowizzard directory the install program tries to replace some dll and I was afraid to break something. Since I use proprietary Mass Spectrometry softwares on my windows computer I did not experiment further because I am not very confident in the way Windows could solve libraries conflicts :-)

**Comment by vdemichev** [comment]
Hi, it should not replace any DLLs, strange that it does. You only need Sciex DLLs for reading .wiff files with DIA-NN, and can also just copy all DLLs with "Clearcore" or "Sciex" in their name from the Proteowizard directory to the folder in which you've installed DIA-NN.

About VCOMP140.DLL, should be solved by installing (maybe reinstall?) the proper Visual C++ redistributable version, it was my mistake, I did not include it in the list of dependencies in the installer, yours is the second report of it missing (in the first case reinstalling the redistributable from this link https://aka.ms/vs/16/release/vc_redist.x64.exe solved the problem). The next DIA-NN version should have this embedded in the installer.

Vadim

**Comment by FredSoftwares** [comment]
Hi Vadim,
Thank you for your advises. I reinstalled Visual C++ and Diann run on windows !
I only had time to do a limited amount of tests. I copied the Sciex and Clearcore DLL but I did not tested yet.

It is possible that I missed or misunderstood something but I did not found in the documentation if Diann can take advantage of spiked iRT (such as Biognosys) and how to make Diann use them. If it is possible to do so, it should be interesting to explain this in the documentation.

Thanks again !
